Ensure that all exited children are reaped properly

--- passt.orig/pasta.c
+++ passt/pasta.c
@@ -87,6 +87,18 @@ void pasta_child_handler(int signal)
 	waitid(P_ALL, 0, NULL, WEXITED | WNOHANG);
 	waitid(P_ALL, 0, NULL, WEXITED | WNOHANG);
 
+	/* Reap all exited children */
+	info("Received SIGCHLD: waiting for children");
+	do {
+		infop.si_pid = 0;
+		if (waitid(P_ALL, 0, &infop, WEXITED | WNOHANG) < 0) {
+			/* waitid returned an error: give up */
+			break;
+		}
+		info("Reaped child %d", infop.si_pid);
+	} while(infop.si_pid != 0);
+	info("Received SIGCHLD: waited for children");
+
 	errno = errno_save;
 }
 
