--- passt.orig/util.h
+++ passt/util.h
@@ -17,6 +17,7 @@
 #include <arpa/inet.h>
 #include <unistd.h>
 #include <sys/syscall.h>
+#include <sys/wait.h>
 
 #include "log.h"
 
@@ -186,10 +187,15 @@ int do_clone(int (*fn)(void *), char *st
 	do {								\
 		char ns_fn_stack[NS_FN_STACK_SIZE]			\
 		__attribute__ ((aligned(__alignof__(max_align_t))));	\
+		int __pid;						\
 									\
-		do_clone((fn), ns_fn_stack, sizeof(ns_fn_stack),	\
+		__pid = do_clone((fn), ns_fn_stack, sizeof(ns_fn_stack),\
 			 CLONE_VM | CLONE_VFORK | CLONE_FILES | SIGCHLD,\
 			 (void *)(arg));				\
+		if (__pid == -1) {					\
+			die("Failed to clone for ns call");		\
+		}							\
+		waitid(P_PID, __pid, NULL, WEXITED);			\
 	} while (0)
 
 #define RCVBUF_BIG		(2ULL * 1024 * 1024)
